File: 135-Candy/solution.py
class Solution(object):
    def candy(self, ratings):
        """
        :type ratings: List[int]
        :rtype: int
        """
        n = len(ratings)
        candys = [1] * n

        for i in range(1, n):
            if ratings[i] > ratings[i - 1]:
                candys[i] = candys[i - 1] + 1

        for i in range(n - 2, -1, -1):
            if ratings[i] > ratings[i + 1] and candys[i] <= candys[i + 1]:
                candys[i] = candys[i + 1] + 1

        return sum(candys)

File: 135-Candy/solution.cpp
class Solution {
public:
    int candy(vector<int>& ratings) {
        vector<int> candys;
        
        candys.resize(ratings.size(), 1);
        
        for (int i = 1; i < (int)ratings.size(); i++) {
            if (ratings[i] > ratings[i - 1]) {
                candys[i] = candys[i - 1] + 1;
            }
        }
        
        for (int i = (int)ratings.size() - 2; i >= 0; i--) {
            if (ratings[i] > ratings[i + 1] && candys[i] <= candys[i + 1]) {
                candys[i] = candys[i + 1] + 1;
            }
        }
        
        int n = 0;
        for (int i = 0; i < (int)candys.size(); i++) {
            n += candys[i];
        }
        
        return n;
    }
};